### ISSUE TITLE
cli: reduce surprise when increasing replication factor for a table/database

### ISSUE THREAD

**Issue by petermattis** [cockroachdb/cockroach#24430]
In https://forum.cockroachlabs.com/t/experiments-on-disaster-recovery/1482/11, the user increased the replication factor on a database to 6, and then tried taking down 3 nodes. This led to data unavailability at which point they brought 1 node back up, thinking the cluster should recover. Unfortunately, they had only increased the replication factor on their database while system ranges contained to be replicated 3 ways.

We should investigate if there are ways to reduce the surprise here. For example, `cockroach zone set` could warn if you try to increase the replication factor on a database/table to a value larger than the replication factor on any system zone (`.meta`, `.liveness` and `system`).

**Comment by bdarnell** [comment]
One reason for confusion is that there are always zone configs set on those system ranges (to reduce GC thresholds), which means that setting the replication factor in `.default` doesn't apply to them. It would be nice if those default zone configs only applied to the GC TTL field and these zones would fall back to the `num_replicas` field in `.default` unless they had had their `num_replicas` changed explicitly.

**Comment by petermattis** [comment]
@bdarnell Yeah, I initially suspected that the user had set the `.default` zone config, but in this case they set a database-level zone config, so your suggestion wouldn't have helped. Might be worthwhile doing what you suggest, but I think something additional is needed.

**Comment by a-robinson** [comment]
The fact that changing `.default` won't affect the replication factor of the zone configs that are prepopulated is a terrible user experience wart that's just asking for trouble. We either need to put a workaround in 2.0.1 (which would be a pretty substantial change in behavior) or document the crap out of it. Currently we don't mention it at all in our docs (cc @jseldess)

**Comment by jseldess** [comment]
OK, I'll add a big warning in a few places for now. Should I also document this as a known limitation? Probably?

**Comment by couchand** [comment]
This just bit another user, _fat0_ from Gitter - https://gitter.im/cockroachdb/cockroach?at=5af1bce05cf0b83004669be1

> I think what was missing is that I was only increasing replication factor for the DB (.default) but not for cluster .liveness and .meta

In this case, the system range fallback to default suggestion would have worked.

As a general rule, it seems like the factor for these two ranges should be at least as high as any other zone.  Would it be crazy to automagically bump it to that value (perhaps warning the user loudly)?

**Comment by a-robinson** [comment]
> Would it be crazy to automagically bump it to that value (perhaps warning the user loudly)?

No, it would not be crazy.

And for future reference when we get to this, the zone config for the `system` database is also important (for table leases and other shared infrastructure).

**Comment by tschottdorf** [comment]
@bdarnell where is this on the cli priority list? This seems pretty bad.

**Comment by bdarnell** [comment]
Not on the roadmap, but it probably should be. We just need to decide which of the three options under discussion we want to use:

- Warn/error when setting a non-system zone config to more replicas than any system one
- Instead of warning, update the system zones to the max of any non-system zone
- Make replication factors adaptive as discussed in https://github.com/cockroachdb/cockroach/issues/14990#issuecomment-391893597

**Comment by knz** [comment]
cc @tbg for re-triage - is this still actionable?

**Comment by tbg** [comment]
The internal zone configs now have a replication factor of 5. If you change the default zone cfg to something higher (say 17) the internal ones will remain at 5. So there's still a problem here though it is reduced in severity since very few users run with a replication factor of higher than five.

**Comment by petermattis** [comment]
@tbg I thought the cascading zone config work paved the way for a real fix for this issue.
